date: 2002/12/30 20:55:34;  author: jer;  state: Exp;  lines: +5 -5
a possible bug relating to blocking sockets and a deadlock, discovered with the help of srlee and the guys over at myjabber

Index: jabberd/mio.c
===================================================================
RCS file: /home/cvs/jabberd14/jabberd/mio.c,v
retrieving revision 1.66
retrieving revision 1.67
diff -u -r1.66 -r1.67
--- jabberd/mio.c	4 Aug 2002 11:30:56 -0000	1.66
+++ jabberd/mio.c	30 Dec 2002 20:55:34 -0000	1.67
@@ -697,11 +697,6 @@
         bind(new->fd, (struct sockaddr*)&sa, sizeof(sa));
     }
 
-    /* set the socket to non-blocking */
-    flags =  fcntl(new->fd, F_GETFL, 0);
-    flags |= O_NONBLOCK;
-    fcntl(new->fd, F_SETFL, flags);
-
 #ifdef WITH_IPV6
     saddr = make_addr_ipv6(cd->ip);
 #else
@@ -743,6 +738,11 @@
         pool_free(p);
         return;
     }
+
+    /* set the socket to non-blocking */
+    flags =  fcntl(new->fd, F_GETFL, 0);
+    flags |= O_NONBLOCK;
+    fcntl(new->fd, F_SETFL, flags);
 
     /* XXX pthreads race condition.. cd->connected may be checked in the timeout, and cd freed before these calls */
 
